### DISCUSSION TITLE
How to translate this piece of code to PascalABC.NET?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#2937] [Q&A] [answered]
```
using System;
using Eto.Forms;
using Eto.Drawing;
using Eto;

namespace Tutorial2
{
	/// <summary>
	/// Custom command
	/// </summary>
	/// <remarks>
	/// You can create your own command subclasses, or create instances of Command directly.
	/// Commands can be used for either the menu or toolbar.
	/// Otherwise, you can use MenuItem or ToolItem classes directly.
	/// </remarks>
	public class MyCommand : Command
	{
		public MyCommand()
		{
			MenuText = "C&lick Me, Command";
			ToolBarText = "Click Me";
			ToolTip = "This shows a dialog for no reason";
			//Image = Icon.FromResource ("MyResourceName.ico");
			//Image = Bitmap.FromResource ("MyResourceName.png");
			Shortcut = Application.Instance.CommonModifier | Keys.M;  // control+M or cmd+M
		}

		protected override void OnExecuted(EventArgs e)
		{
			base.OnExecuted(e);

			MessageBox.Show(Application.Instance.MainForm, "You clicked me!", "Tutorial 2", MessageBoxButtons.OK);
		}
	}

	public class MyForm : Form
	{
		public MyForm()
		{
			ClientSize = new Size(600, 400);
			Title = "Menus and Toolbars";

			// create menu
			Menu = new MenuBar
			{
				Items =
				{
					new ButtonMenuItem
					{ 
						Text = "&File",
						Items =
						{ 
							// you can add commands or menu items
							new MyCommand(),
							new ButtonMenuItem { Text = "Click Me, MenuItem" }
						}
					} 
				},
				// quit item (goes in Application menu on OS X, File menu for others)
				QuitItem = new Command((sender, e) => Application.Instance.Quit())
				{ 
					MenuText = "Quit", 
					Shortcut = Application.Instance.CommonModifier | Keys.Q
				},
				// about command (goes in Application menu on OS X, Help menu for others)
				AboutItem = new Command((sender, e) => new Dialog { Content = new Label { Text = "About my app..." }, ClientSize = new Size(200, 200) }.ShowModal(this))
				{ 
					MenuText = "About my app"
				}
			};

			// create toolbar
			ToolBar = new ToolBar
			{
				Items =
				{ 
					new MyCommand(),
					new SeparatorToolItem(),
					new ButtonToolItem { Text = "Click Me, ToolItem" }
				}
			};
		}
	}

	class Program
	{
		[STAThread]
		public static void Main(string[] args)
		{
			new Application().Run(new MyForm());
		}
	}
}
```

My partial translation:

```
{$apptype windows}

uses 
  Eto.Forms,
  Eto.Drawing,
  Eto;

type
  MyCommand = class(Command)
  public
    constructor;
    begin
      MenuText := 'C&lick Me, Command';
			ToolBarText := 'Click Me';
			ToolTip := 'This shows a dialog for no reason';
			Shortcut := Application.Instance.CommonModifier or Keys.M;
    end;
    procedure OnExecuted(e: System.EventArgs); override;
    begin
      //base.OnExecuted(e); //Project4.pas(20) : Undefined name reference 'base'
      MessageBox.Show(Application.Instance.MainForm, 'You clicked me!', 'Tutorial 2', MessageBoxButtons.OK);
    end;
  end;
  
  MyForm = class(Form)
  public
    constructor;
    begin
      Title := 'Menus and Toolbars';
      ClientSize:= new Eto.Drawing.Size(600, 400);
      
      var Menu := new &MenuBar;
      // don't know how to translate, too complex!
    end;
  end;

begin
  (new Application).Run(new MyForm);
end.
```

**Comment by ghost** [comment]
Update: I'm currently replace `base` with `self` and it compiled. I don't know if this is right or not since as I know, `self` is equivalent to `this`. The C# code I'm not translated is still a problem, though. Please help. Thank you.

**Reply by SunSerega** [reply]
Instead of a `base` variable, you use the `inherited` keyword before the name you are referencing.
So `base.Func(a)` would be `inherited Func()`.
And `base.Prop.Func(a)` would be `(inherited Prop).Func()`. In most cases, it would also work as `inherited Prop.Func()`.
Well, in this case, you need neither, since your own class doesn't define anything that would cover up the `inherited` method. It's just a good practice to be direct when you code. It may look useless while writing, but it sometimes saves you much more time later on.

**Reply by SunSerega** [reply]
> don't know how to translate, too complex!

I'm pretty sure you do, and you don't like it)) Well, as an alternative you can actually load XAML at run time. I know there are WPF functions for that - and `Eto.Serialization.Xaml` looks like it exists as an equivalent for Eto. Not sure if it would actually make it easier in the end tho...

**Reply by SunSerega** [reply]
> `(new Application).Run(new MyForm);`

```
Application.Create.Run(new MyForm);
```
Older syntax, but it looks better here. Fewer brackets = easier on the eyes.

**Reply by ghost** [reply, marked as answer]
Any idea to shorten the code?

```
{$apptype windows}

uses 
  Eto.Forms,
  Eto.Drawing,
  Eto;

type
  MyCommand = class(Command)
  public
    constructor;
    begin
      MenuText := 'C&lick Me, Command';
      ToolBarText := 'Click Me';
      ToolTip := 'This shows a dialog for no reason';
      Shortcut := Application.Instance.CommonModifier or Keys.M;
    end;
    
    procedure OnExecuted(e: System.EventArgs); override;
    begin
      inherited OnExecuted(e);
      MessageBox.Show(Application.Instance.MainForm, 'You clicked me!', 'Tutorial 2', MessageBoxButtons.OK);
    end;
  end;
  
  MyForm = class(Form)
  public
    constructor;
    begin
      Title := 'Menus and Toolbars';
      ClientSize := new Eto.Drawing.Size(600, 400);
      
      Menu := new Eto.Forms.MenuBar;
      var btnMenuItem := new ButtonMenuItem;
      btnMenuItem.Text := '&File';
      var btnSubMenuItem := new ButtonMenuItem;
      btnSubMenuItem.Text := 'Click Me, MenuItem';
      var myCmd := new MyCommand;
      btnMenuItem.Items.Add(myCmd);
      btnMenuItem.Items.Add(btnSubMenuItem);
      Menu.Items.Add(btnMenuItem);
      Menu.QuitItem := new Command((sender, e) -> Application.Instance.Quit());
      Menu.QuitItem.Text := 'Quit';
      Menu.QuitItem.Shortcut := Application.Instance.CommonModifier or Keys.Q;
      Menu.AboutItem := new Command((sender, e) -> begin
        var MyDialog := new Dialog;
        var lbl1 := new &Label;
        lbl1.Text := 'About my app...';
        MyDialog.Content := lbl1;
        MyDialog.ClientSize := new Eto.Drawing.Size(200, 200);
        MyDialog.ShowModal();
      end);
      Menu.AboutItem.Text := 'About my app';
      ToolBar := new Eto.Forms.ToolBar;
      ToolBar.Items.Add(new MyCommand);
      ToolBar.Items.Add(new SeparatorToolItem);
      var btnToolItem := new ButtonToolItem;
      btnToolItem.Text := 'Click Me, ToolItem';
      ToolBar.Items.Add(btnToolItem);
    end;
  end;

begin
  (new Application).Run(new MyForm);
end.
```

It's better if I could assign `Items` directly or at least use `AddRange` (disclaimer: I don't really understand what this method does and how to use it, I only guess from the name of it). Using the `Add` method to add each item manually is too verbose.

**Reply by SunSerega** [reply]
Well, only you see what `AddRange` accepts. But I would guess it's `IEnumerable<something>` (can also be written as `sequence of something`). In that case, you can just pass in an array - it is the simplest implementation of `IEnumerable`.
```
.AddRange(|
  item1,
  item2,
  item3
|);
```

**Reply by SunSerega** [reply]
> Any idea to shorten the code?

Would probably not shorten it, but I can see a possibility of creating make-shift ObjectInitializers using a helper function. Something like:
```
function CreateAndInit<T>(params inits: array of T->()): T;
  where T: constructor;
begin
  Result := new T;
  foreach var init in inits do init(Result);
end;
```
And then use it similar to this:
```
.Content := CreateAndInit&<ControlType>(
  x->(x.Prop1 := val1),
  x->(x.Prop2 := CreateAndInit)
);
```
The key point is that you can stack calls to `CreateAndInit` inside each other.

**Reply by ghost** [reply]
You are right about which inputs `AddRange` expects. But the code you give me has compilation error. `MyCmd` is the one problematic here. `AddRange` expects `sequence of MenuItem` but `MyCmd` is of type `MyCommand`.

**Reply by ghost** [reply]
Here is the declaration of `Items` by using the `Go To Declaration` feature of the IDE:

```
    /// Gets the collection of menu items.
    public property Items : Eto.Forms.MenuItemCollection read;
```

In ILSpy, `MenuItemCollection` is declared as:

`public class MenuItemCollection : Collection<MenuItem>, IList, ICollection, IEnumerable`

**Reply by ghost** [reply]
`Items` is a readonly property so I can't assign to it. I can only use `AddRange`.

**Reply by SunSerega** [reply]
Yes, you are not supposed to assign to properties like `.Items`. Doing that actually means clearing previous items and adding a range of new items.

> `AddRange` expects `sequence of MenuItem` but `MyCmd` is of type `MyCommand`.

Does `MyCommand` inherit from `MenuItem`? In that case, you can explicitly specify the array item type:
```
.AddRange(new MenuItem[](
  item1,
  item2,
  item3
));
```
But that doesn't sound right to me... You probably need to create a new object of type `MenuItem` and assign `MyCmd` to one of its properties.

**Reply by ghost** [reply]
`MyCommand` inherits from `Command` and in ILSpy `Command` is declared as:

`public class Command : IBindable, ICommand`

**Reply by SunSerega** [reply]
Hmm, looking at the original C# code - I guess there is some kind of `operator implicit` defined in Eto, to automatically convert commands to very basic menu items.

**Reply by ghost** [reply]
It seems there is no way other than adding each item manually.
